tests/aws-cpp-sdk-core-tests/aws/auth/CredentialTrackingTest.cpp

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/testing/AwsCppSdkGTestSuite.h>
+#include <aws/testing/AwsTestHelpers.h>
+#include <aws/testing/mocks/aws/client/MockAWSClient.h>
+#include <aws/testing/mocks/http/MockHttpClient.h>
+#include <aws/core/auth/AWSCredentialsProvider.h>
+#include <aws/core/client/AWSClient.h>
+#include <aws/core/utils/StringUtils.h>
+#include <aws/core/platform/Environment.h>
+
+using namespace Aws::Client;
+using namespace Aws::Auth;
+using namespace Aws::Http;
+
+static const char ALLOCATION_TAG[] = "CredentialTrackingTest";
+
+
+
+class CredentialTrackingTest : public Aws::Testing::AwsCppSdkGTestSuite
+{
+protected:
+    std::shared_ptr<MockHttpClient> mockHttpClient;
+    std::shared_ptr<MockHttpClientFactory> mockHttpClientFactory;
+
+    void SetUp() override
+    {
+        mockHttpClient = Aws::MakeShared<MockHttpClient>(ALLOCATION_TAG);
+        mockHttpClientFactory = Aws::MakeShared<MockHttpClientFactory>(ALLOCATION_TAG);
+        mockHttpClientFactory->SetClient(mockHttpClient);
+        SetHttpClientFactory(mockHttpClientFactory);
+    }
+
+    void TearDown() override
+    {
+        mockHttpClient->Reset();
+        mockHttpClient = nullptr;
+        mockHttpClientFactory = nullptr;
+        Aws::Http::CleanupHttp();
+        Aws::Http::InitHttp();
+    }
+};
+
+TEST_F(CredentialTrackingTest, TestEnvironmentCredentialsTracking)
+{
+    Aws::Environment::SetEnv("AWS_ACCESS_KEY_ID", "test-access-key", 1);
```

Comment:
i made a utility for this called [`EnvironmentRAII`](https://github.com/aws/aws-sdk-cpp/blob/main/tests/testing-resources/include/aws/testing/platform/PlatformTesting.h#L28-L42) that is a helper function to set/unset env vars in tests so this could be.
```cpp
Aws::Environment::EnvironmentRAII testEnvironment{{
  {"AWS_ACCESS_KEY_ID", "test-access-key"},
  {"AWS_SECRET_ACCESS_KEY", "test-secret-key"},
}};
```

then you dont need to worry about setting/unsetting env vars
```